## Exploring Hackers News Posts
In this project, I'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts I'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

I'll specifically compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

It should be noted that the data set I am working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

## Introduction
First, I'll read in the data and remove the headers.

In [1]:
import csv
file = open('hacker_news.csv')
hn = list(csv.reader(file))
hn = hn[1:]

## Extracting Ask HN and Show HN Posts
First, I'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [2]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
        

## Calculating the Average Number of Comments for Ask HN and Show HN Posts

Now that I have separated ask posts and show posts into different lists, I'll calculate the average number of comments each type of post receives

In [3]:
total_ask_comments = 0

for post in ask_posts:
    comments = int(post[4])
    total_ask_comments += comments
avg_ask_comments = total_ask_comments / len(ask_posts)

print(avg_ask_comments)

14.038417431192661


In [4]:
total_show_comments = 0

for post in show_posts:
    comments = int(post[4])
    total_show_comments += comments
avg_show_comments = total_show_comments / len(show_posts)

print(avg_show_comments)

10.31669535283993


On average, ask posts in sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, I'll focus my remaining analysis just on these posts.

In [9]:
import datetime as dt
result_list = []
for post in ask_posts:
    result_list.append([dt.datetime.strptime(post[6], "%m/%d/%Y %H:%M")
    ,int(post[4])])
    
comments_by_hour = {}
counts_by_hour = {}

for row in result_list:
    date = row[0]
    comment = row[1]
    time = date.strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1
print(comments_by_hour) 

{'08': 492, '13': 1253, '07': 267, '05': 464, '10': 793, '17': 1146, '19': 1188, '00': 447, '02': 1381, '03': 421, '09': 251, '20': 1722, '11': 641, '22': 479, '12': 687, '06': 397, '23': 543, '04': 337, '16': 1814, '18': 1439, '21': 1745, '01': 683, '14': 1416, '15': 4477}


## Calculating the Average Number of Comments for Ask HN Posts by Hour ¶

In [12]:
avg_by_hour = []
for row in comments_by_hour:
    avg_by_hour.append([row,comments_by_hour[row]/ counts_by_hour[row]])

## Sorting and Printing Values from a List of Lists

In [15]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[0],row[1]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("\n ----------------------------------------      \n")
print(sorted_swap)

[['08', 10.25], ['13', 14.741176470588234], ['07', 7.852941176470588], ['05', 10.08695652173913], ['10', 13.440677966101696], ['17', 11.46], ['19', 10.8], ['00', 8.127272727272727], ['02', 23.810344827586206], ['03', 7.796296296296297], ['09', 5.5777777777777775], ['20', 21.525], ['11', 11.051724137931034], ['22', 6.746478873239437], ['12', 9.41095890410959], ['06', 9.022727272727273], ['23', 7.985294117647059], ['04', 7.170212765957447], ['16', 16.796296296296298], ['18', 13.20183486238532], ['21', 16.009174311926607], ['01', 11.383333333333333], ['14', 13.233644859813085], ['15', 38.5948275862069]]

 ----------------------------------------      

[['23', 7.985294117647059], ['22', 6.746478873239437], ['21', 16.009174311926607], ['20', 21.525], ['19', 10.8], ['18', 13.20183486238532], ['17', 11.46], ['16', 16.796296296296298], ['15', 38.5948275862069], ['14', 13.233644859813085], ['13', 14.741176470588234], ['12', 9.41095890410959], ['11', 11.051724137931034], ['10', 13.4406779661016

In [ ]:
print("TOP 5  Hours for ASK Posts Comments")

for avg,hr in sorted_swap:
    print("{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.